app/models/user.rb

```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+
+  # Other Devise configurations
+
+  validate :password_complexity, if: :password_required?
+
+  def password_complexity
+    return if password.blank? || password.match?(/(?=.*[A-Z])(?=.*[a-z])(?=.*\d).{8,}/)
+
+    errors.add(:password, 'must include at least one uppercase letter, one lowercase letter, and one digit')
+  end
```

Comment:
As we've already checked it with `password_required?`, do you think we need this check? 
Also, we can define the PASSWORD_PATTERN as a constant.
Moreover, we can add a custom `type` for the `error`, and then we can [add the locale for this error](https://guides.rubyonrails.org/v3.2.0/i18n.html#:~:text=For%20example%2C%20you%20might%20have%20an%20Admin%20model%20inheriting%20from%20User%3A) later.

```suggestion
  PASSWORD_PATTERN = /(?=.*[A-Z])(?=.*[a-z])(?=.*\d).{8,}/
  
  def password_complexity
    return if password.match?(PASSWORD_PATTERN)

    errors.add(:password, :invalid)
  end
```

Reply:
Thanks for this great suggestion 👍🏼 . I tried to add this extra checking for password just to see how it's done in rails philosophy. More dirty hands, more learning 😄. Now going with your suggested `:invalid` message. I'll try to add custom validation message later.